### PR TITLE
Search widgets

### PR DESCRIPTION
**Pull request by carolinux**

### CODE REVIEW

File: src/app/qgsattributetabledialog.cpp [hunk 1]
```diff
@@ -706,39 +710,15 @@ void QgsAttributeTableDialog::filterQueryChanged( const QString& query )
...
-    const QgsFields& flds = mLayer->pendingFields();
-    int fldIndex = mLayer->fieldNameIndex( fieldName );
-    if ( fldIndex < 0 )
-      return;
 
-    QVariant::Type fldType = flds[fldIndex].type();
-    bool numeric = ( fldType == QVariant::Int || fldType == QVariant::Double || fldType == QVariant::LongLong );
-
-    QString sensString = "ILIKE";
+    QString sensString = "ILIKE"; // FIXME need to take this into account
```
Comment (m-kuhn) [inline src/app/qgsattributetabledialog.cpp]: This should probably go into the QgsDefaultSearchWidget (it's unrequired for the comboboxes). You can make a QWidget as container and nest a QLineEdit and the checkbox in it.

File: src/app/qgsattributetabledialog.cpp [hunk 2]
```diff
@@ -435,6 +435,10 @@ void QgsAttributeTableDialog::filterColumnChanged( QObject* filterAction )
   const QgsEditorWidgetConfig widgetConfig = mLayer->editorWidgetV2Config( fldIdx );
   mCurrentSearchWidgetWrapper = QgsEditorWidgetRegistry::instance()->
                                 createSearchWidget( widgetType, mLayer, fldIdx, widgetConfig, mFilterContainer );
+  if (mCurrentSearchWidgetWrapper->applyDirectly())
```
Comment (m-kuhn) [inline src/app/qgsattributetabledialog.cpp]: I would remove this condition and just connect it always.
If you are afraid that there is a performance problem / too many events from a search widget, it should be fixed inside the search widget. So i.e. having an apply button or not should be the widget's responsibility.

File: src/app/qgsattributetabledialog.cpp (hunk 1 above)
Comment (carolinux) [inline src/app/qgsattributetabledialog.cpp]: uggh then I need to change the layout ui again. but it makes sense conceptually for the searchwidget to contain the UI elems that define its behavior!

Comment (m-kuhn) [inline src/app/qgsattributetabledialog.cpp]: That should be an easy one :)

File: src/app/qgsattributetabledialog.cpp (hunk 2 above)
Comment (carolinux) [inline src/app/qgsattributetabledialog.cpp]: ok

Comment (m-kuhn) [inline src/app/qgsattributetabledialog.cpp]: Although... Maybe you can leave it. I just thought that in the future there could be several on a separate search dialog to build complex queries. If every of these brings it's own apply button it's a bit strange.

Comment (carolinux) [inline src/app/qgsattributetabledialog.cpp]: They dont have their own apply button, it's just that you can define whether you want their changes to take effect immediately or not (not very good decoupling here between attribute table and search widget)

**Comment by m-kuhn** [comment]
Hey Karolina,

I think it's almost there :)
Did you make any progress?

**Comment by carolinux** [comment]
You commented just when I started working on it again.

**Comment by m-kuhn** [comment]
That's always good to hear :)

**Comment by carolinux** [comment]
What should I use as a container? QWidget somehow doesnt show anything when im like
```
qwidget = Qwidget(parent)
mLineEdit = QgsLineFilter(qwidget)
```

**Comment by m-kuhn** [comment]
QWidget can be used as container:

```
QWidget* w = new QWidget();
w->setLayout(new QGridLayout() );
w->layout()->addWidget( new QgsLineFilter() );
```

N.B. widgets get parented automatically when added to another widget.

**Comment by carolinux** [comment]
got it to work thanks. should probably be a horizontal layout though :)

**Comment by m-kuhn** [comment]
whatever suits your needs best :)

**Comment by carolinux** [comment]
Alright, so I think now what is left is to create widgets for some more types (valuerelation, anything else?).  About something you said in an email, I think the inheritance is desirable because of the shared nice initialization functions from QgsWidgetWrapper

**Comment by m-kuhn** [comment]
Cool!
ValueRelation is the one that should be implemented for 2.10, RelationReference for 2.12 I guess.

Let's keep the QgsWidgetWrapper in this case. It has a bit more than required (e.g. the context) and the necessary methods could be factored out to a common base class. But it won't hurt and maybe there will be use for the context in the future (e.g. searching for fields from a related layer and embedding a search form...)

**Comment by carolinux** [comment]
valuerelationwidget wrapper has a lot of code.. I dont want to duplicate it, should I just subclass it for the search widget?

**Comment by carolinux** [comment]
Not sure of the signifinace of this
```
  if ( config( "AllowMulti" ).toBool() )
  {
    return new QListWidget( parent );
  }
  else if ( config( "UseCompleter" ).toBool() )
  {
    return new QgsFilterLineEdit( parent );
  }
  {
    return new QComboBox( parent );
```

**Comment by carolinux** [comment]
kinda blocked on this, not sure what the best way is to proceed with the value relation.. copy paste for now?

**Comment by m-kuhn** [comment]
For the multitype, can you inherit from the default widget and return a lineedit in this case? Proper support for it is probably not straightforward. (I.e. postponed to 2.12)

For the shared code: I think that should go into a separate helper class. If you copy paste it for now during feature freeze I don't have any objections.

**Comment by carolinux** [comment]
What do you mean "multitype" - the allowMulti case?

**Comment by m-kuhn** [comment]
Yes :) sorry

**Comment by m-kuhn** [comment]
Btw, I think it would be very easy to fix this issue while working on this code anyway:

http://hub.qgis.org/issues/11627

It's just a matter of using case-insensitive search as default in the defaultsearchwidget

**Comment by carolinux** [comment]
ok, sure.. I thought case sensitive as default was a good idea fwiw

**Comment by m-kuhn** [comment]
I think realizing that you need to switch to case sensitive mode if there are too many results is obvious. But if there are no results shown at all and you don't know why can be quite frustrating.

**Comment by carolinux** [comment]
yes for this I just need to delete one line

**Comment by m-kuhn** [comment]
Nice!

**Comment by carolinux** [comment]
Can you share a project file with a valuerelationmap built in ? For testing.

**Comment by m-kuhn** [comment]
I don't have anything ready, but it's easy.

You can create a CSV with

1, One
2, Two
3, Three

And load it as (geometryless) layer

And another e.g. shapefile with a field "test" of type integer and define the widget in layer properties -> fields as valuerelation and configure it to look up the values on the CSV layer.

**Comment by carolinux** [comment]
OK, I set this up. One problem I found is that the combobox has selected the first value. This doesn't work right now because no "expressionChanged" is emitted. Also it doesn't make sense to select the first value automatically. Perhaps all comboboxes should have a " --All features--" item?

**Comment by m-kuhn** [comment]
Cool, looking forward to testing it.

Yes, I think an "all features" or "please select" item makes sense

**Comment by carolinux** [comment]
I think now it's ready for testing.

**Pull request closed without merging**